sw/device/lib/dif/dif_pinmux.h

```diff
@@ -0,0 +1,656 @@
+   * The "disabled" state.
+   */
+  kDifPinmuxToggleDisabled,
+} dif_pinmux_toggle_t;
+
+/**
+ * Hardware instantiation parameters for a Pin Multiplexer.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of a top-level
+ * design.
+ */
+typedef struct dif_pinmux_params {
+  /**
+   * The base address for the a Pin Multiplexer hardware registers.
+   */
+  mmio_region_t base_addr;
+} dif_pinmux_params_t;
+
+/**
+ * A handle to a Pin Multiplexer.
+ *
+ * This type should be treated as opaque by users.
+ */
+typedef struct dif_pinmux {
+  dif_pinmux_params_t params;
+} dif_pinmux_t;
+
+/**
+ * Pin Multiplexer Padring pad kinds.
+ *
+ * A pad can be a Multiplexed (MIO) or Dedicated (DIO).
+ */
+typedef enum dif_pinmux_pad_kind {
+  /**
+   * Multiplexed Input/Output pad. Connected to a peripheral input or output
+   * via a multiplexer matrix inside the a Pin Multiplexer.
+   */
+  kDifPinmuxPadKindMio = 0,
+  /**
+   * Dedicated Input/Output pad. Connected directly to a peripheral input
+   * or output, bypassing the multiplexer matrix.
+   */
+  kDifPinmuxPadKindDio,
+} dif_pinmux_pad_kind_t;
+
+/**
+ * Pin Multiplexer functionality locking.
+ *
+ * Represents a target functionality to be locked for a given
+ * pad/peripheral input.
+ */
+typedef enum dif_pinmux_lock_target {
+  /**
+   * Lock MIO pad input select for a given peripheral input. This means that
+   * the peripheral input connection can no longer be changed.
+   */
+  kDifPinmuxLockTargetInsel = 0,
+  /**
+   * Lock peripheral output select for a given MIO pad output. This means
+   * that the MIO pad output connection can no longer be changed.
+   */
+  kDifPinmuxLockTargetOutsel,
+  /**
+   * Lock the sleep mode configuration for a given MIO pad.
+   */
+  kDifPinmuxLockTargetMioSleep,
+  /**
+   * Lock the sleep mode configuration for a given DIO pad.
+   */
+  kDifPinmuxLockTargetDioSleep,
+  /**
+   * Lock physical characteristics configuration for a given MIO pad.
+   */
+  kDifPinmuxLockTargetMioPadAttr,
+  /**
+   * Lock physical characteristics configuration for a given DIO pad.
+   */
+  kDifPinmuxLockTargetDioPadAttr,
+  /**
+   * Lock wake-up detector configuration for a given detector.
+   */
+  kDifPinmuxLockTargetWakeupDetector,
+
+} dif_pinmux_lock_target_t;
+
+/**
+ * Pin Multiplexer generic index.
+ *
+ * The meaning of this index changes between APIs, and must be documented in
+ * the corresponding function description.
+ */
+typedef uint32_t dif_pinmux_index_t;
+
+/**
+ * Pin multiplexer Padring pad attributes.
+ *
+ * This `enum` is a bitfield, meaning that it is allowed to pass combined value
+ * of multiple individual attributes.
+ *
+ * Some attributes have defined meanings, others are implementation defined. Not
+ * all pads will support all attributes, or attribute combinations. The Pin
+ * Multiplexer DIF does not check for conflicting or unsupported attributes,
```

Comment:
> The Pin Multiplexer DIF does not check for conflicting or unsupported attributes

This appears to contradict the comment on `dif_pinmux_pad_write_attrs`

> Not all pads implement all attributes and some combinations cannot be enabled together. This function returns  a `kDifPinmuxPadWriteAttrsConflict` error in case of an invalid `attrs_in`. Conflicting attributes will be discarded by the hardware, and can be identified by comparing `attrs_in` to `attrs_out`.

I think the idea is the DIF just writes the attributes to the register and then reads back. You can determine you did something unsupported by the read back value not matching the written value but looks like it's up to the API user to do this? If the DIF isn't doing any checking (either explicit conflict checks or using WARL behaviour) what is `kDifPinmuxPadWriteAttrsConflict` for?

Reply:
I suspect the documentation is slightly off, so the behaviour now is - any function that writes to the ATTRS register will read back the value (WARL) and error `kDifPinmuxPadWriteAttrsConflict` if there is a conflict. We had a quick discussion with the SW people and agreed that we think this way is better. It is however different from Sam (I believe) was suggesting in the original Padctrl review.